### PR TITLE
Kyv 171 offboard route route options 

### PR DESCRIPTION
**Pull request by RingerJK**
## Description

Please include a brief summary of the change and which issue is fixed (e.g., fixes [#issue](link))

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Please describe the PR goals. Just the stuff needed to implement the fix / feature and a simple rationale. It could contain many check points if needed

### Implementation

Please include all the relevant things implemented and also rationale, clarifications / disclaimers etc. related to the approach used. It could be as self code companion comments

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

**Comment by Guardiola31337** [comment]
@RingerJK 👀 https://github.com/mapbox/mapbox-navigation-android/pull/2286/commits/928fa0d145d9228fe3a03c7d6a95d7905d9cabac and let me know what you think.

I also captured all the parameters in `MapboxDirections` / `NavigationRoute` in the following table:

| `RouteOptionsNavigation` | Default (`MapboxDirections`)              | Default (`NavigationRoute`)                                                            |                      | Mandatory / Optional |
| ------------------------ | ----------------------------------------- | -------------------------------------------------------------------------------------- | -------------------- | -------------------- |
| `baseUrl`                | `Constants.BASE_API_URL`                  |                                                                                        |                      | M                    |
| `user`                   | `DirectionsCriteria.PROFILE_DEFAULT_USER` |                                                                                        |                      | ?                    |
| `profile`                | `DirectionsCriteria.PROFILE_DRIVING`      | `DirectionsCriteria.PROFILE_DRIVING_TRAFFIC`                                           |                      | M                    |
| `coordinates`            |                                           |                                                                                        |                      | M                    |
| `alternatives`           |                                           |                                                                                        |                      | O                    |
| `language`               |                                           | `language(Context)`                                                                    |                      | O                    |
| `radiuses`               |                                           |                                                                                        |                      | O                    |
| `bearings`               |                                           |                                                                                        |                      | O                    |
| `continueStraight`       |                                           | `true`                                                                                 |                      | O                    |
| `roundaboutExits`        |                                           | `true`                                                                                 |                      | O                    |
| `geometries`             | `DirectionsCriteria.GEOMETRY_POLYLINE6`   | `DirectionsCriteria.GEOMETRY_POLYLINE6`                                                |                      | O                    |
| `overview`               |                                           | `DirectionsCriteria.OVERVIEW_FULL`                                                     |                      | O                    |
| `steps`                  |                                           | `true`                                                                                 |                      | O                    |
| `annotations`            |                                           | `DirectionsCriteria.ANNOTATION_CONGESTION`<br>`DirectionsCriteria.ANNOTATION_DISTANCE` |                      | O                    |
| `voiceInstructions`      |                                           | `true`                                                                                 |                      | O                    |
| `bannerInstructions`     |                                           | `true`                                                                                 |                      | O                    |
| `voiceUnits`             |                                           | `voiceUnits(Context)`                                                                  |                      | O                    |
| `accessToken`            |                                           |                                                                                        |                      | M                    |
| `requestUuid`            |                                           |                                                                                        | Set at response time | O                    |
| `exclude`                |                                           |                                                                                        |                      | O                    |
| `approaches`             |                                           |                                                                                        |                      | O                    |
| `waypointIndices`        |                                           |                                                                                        |                      | O                    |
| `waypointNames`          |                                           |                                                                                        |                      | O                    |
| `waypointTargets`        |                                           |                                                                                        |                      | O                    |
| `walkingOptions`         |                                           |                                                                                        |                      | O                    |
|                          |                                           |                                                                                        |                      |                      |
| `clientAppName`          | X                                         | X                                                                                      | Never set internally | ?                    |
| `interceptor`            | X                                         | X                                                                                      |                      | ?                    |
| `eventListener`          | X                                         | `NavigationRouteEventListener`                                                         | Internal?            | ?                    |
| `enableRefresh`          | X                                         | `true`                                                                                 |                      | O                    |
| `post`                   | X                                         |                                                                                        |                      | M                    |
| `get`                    | X                                         |                                                                                        |                      | M                    |

Based on ☝️ we should review the default parameters + if they're mandatory or optional and adapt `NavigationRoute`, `RouteOptionsNavigation`, `Mappers` and types nullability accordingly.

I also left some `TODO`s.

**Comment by RingerJK** [comment]
I think that we must have only one source of truth(`RouteOptionsNavigation` fits imo). This source will have own default values(we choose by our meaning). So we could skip this table and have only one with `RouteOptionsNavigation` default values. wdyt? **cc** @LukasPaczos

**Comment by Guardiola31337** [comment]
> Based on ☝️ we should review the default parameters + if they're mandatory or optional and adapt `NavigationRoute`, `RouteOptionsNavigation`, `Mappers` and types nullability accordingly.

> I think that we must have only one source of truth(`RouteOptionsNavigation` fits imo). This source will have own default values(we choose by our meaning). So we could skip this table and have only one with `RouteOptionsNavigation` default values. wdyt?

Per discussion, we're going to address these concerns as part of [`Onboard` PR](https://github.com/mapbox/mapbox-navigation-android/pull/2253) and other separate ones - as these changes shouldn't affect public `interface`s and will unblock merging https://github.com/mapbox/mapbox-navigation-android/pull/2240

Great work @RingerJK 💪 🚀

File: libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouterTest.kt [hunk 1]
```diff
@@ -19,26 +19,22 @@ import retrofit2.Response
 
 class MapboxOffboardRouterTest : BaseTest() {
 
-    companion object {
-        private const val ACCESS_TOKEN = "pk.token"
-    }
-
     private val navigationRoute = mockk<NavigationRoute>(relaxed = true)
+    // private val navigationRoute = NavigationRoute.builder(mockk(relaxed = true)).build()
```
Comment (Guardiola31337) [inline libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouterTest.kt]: NIT Should we remove this commented out code?

File: libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/router/NavigationRouteTest.kt [hunk 2]
```diff
@@ -129,10 +145,18 @@ class NavigationRouteTest {
...
+            .routeOptions(
+                RouteOptionsNavigation.builder().accessToken(ACESS_TOKEN)
+                    .origin(origin)
+                    .destination(destination)
+                    .waypointTargets(
+                        arrayOf(
+                            Point.fromLngLat(0.99, 4.99),
+                            Point.fromLngLat(1.99, 5.99)
+                        )
+                            .joinToString(separator = ";") { "${it?.latitude()},${it?.longitude()}" }
```
Comment (Guardiola31337) [inline libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/router/NavigationRouteTest.kt]: NIT no new line needed

```suggestion
).joinToString(separator = ";") { "${it?.latitude()},${it?.longitude()}" }
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/DirectionsSession.kt [hunk 3]
```diff
@@ -1,15 +1,13 @@
...
 import com.mapbox.navigation.base.route.model.Route
+import com.mapbox.navigation.base.route.model.RouteOptionsNavigation
 
 interface DirectionsSession {
     fun getRoutes(): List<Route>
-    fun getOrigin(): Point?
-    fun getWaypoints(): List<Point>
-    fun getDestination(): Point?
+    fun getRouteOptions(): RouteOptionsNavigation?
     // note: write in javadoc for java devs about default param waypoints
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/route/DirectionsSession.kt]: NIT This comment is out of date and should be removed

File: libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouterTest.kt (hunk 1 above)
Comment (RingerJK) [reply libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouterTest.kt]: removed

File: libdirections-offboard/src/main/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouter.kt [hunk 4]
```diff
@@ -14,52 +13,21 @@ import retrofit2.Callback
 import retrofit2.Response
 
 @MapboxNavigationModule(MapboxNavigationModuleType.OffboardRouter, skipConfiguration = true)
-class MapboxOffboardRouter : Router {
+class MapboxOffboardRouter(
+    private val routeBuilderProvider: NavigationRoute.Builder
```
Comment (Lebedinsky) [inline libdirections-offboard/src/main/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouter.kt]: Should we rename it to routeBuilder?

File: libnavigation-core/build.gradle [hunk 5]
```diff
@@ -29,6 +29,7 @@ dependencies {
     implementation project(':libdirections-session')
     implementation project(':libtrip-session')
     implementation(project(':libnavigator'))
+    implementation(project(':libdirections-offboard'))
```
Comment (Lebedinsky) [inline libnavigation-core/build.gradle]: Should `core` depend on `offboard` ?

Comment (Guardiola31337) [reply libnavigation-core/build.gradle]: This "core" module is what will become `libandroid-navigation` `1.0`, so will likely include all of the modules. We created a separate module for early integration and testing.

File: libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouterTest.kt (hunk 1 above)
Comment (Guardiola31337) [reply libdirections-offboard/src/test/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouterTest.kt]: Was marked as resolved but the comment is still here. Did you push :trollface:?

**Pull request merged**